File: Noteably-main/backend/noteably/src/main/java/com/g3appdev/noteably/Service/TimerService.java
package com.g3appdev.noteably.Service;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.g3appdev.noteably.Entity.TimerEntity;
import com.g3appdev.noteably.Repository.TimerRepository;

@Service
public class TimerService {

    @Autowired
    private TimerRepository timerRepo;

    public TimerEntity createTimer(TimerEntity timer) { 
        // Set the studentId for the timer
        if (timer.getStudentId() == 0) {
            throw new IllegalArgumentException("Student ID must be provided");
        }
        return timerRepo.save(timer);
    }
    
    public List<TimerEntity> getTimersByStudentId(int studentId) {
        return timerRepo.findByStudentId(studentId);
    }

    public List<TimerEntity> getAllTimers() {
        return timerRepo.findAll();
    }

    public TimerEntity getTimerById(int id) {
        return timerRepo.findById(id).orElse(null);
    }

    public TimerEntity updateTimer(int id, TimerEntity newTimerDetails) {
        TimerEntity timer = timerRepo.findById(id).orElse(null);
        if (timer != null) {
            timer.setTitle(newTimerDetails.getTitle());
            timer.setHours(newTimerDetails.getHours());
            timer.setMinutes(newTimerDetails.getMinutes());
            timer.setSeconds(newTimerDetails.getSeconds());
            return timerRepo.save(timer);
        }
        return null;
    }

    public String deleteTimer(int id) {
        timerRepo.deleteById(id);
        return "Timer with ID: " + id + " deleted.";
    }

    public void startTimer(int timerID) {
        TimerEntity timer = getTimerById(timerID);
        if (timer != null) {
            // Logic to start the timer (e.g., set a status, schedule a job, etc.)
            // This is a placeholder for actual timer logic.
            // You may want to implement a scheduling mechanism or a background job.
            System.out.println("Timer started: " + timer.getTitle());
        } else {
            throw new RuntimeException("Timer not found");
        }
    }
}

File: Noteably-main/frontend/noteably/src/services/timerService.js
import { axiosInstance } from '../studentService';

const BASE_URL = 'http://localhost:8080/api/timer';

const timerService = {
    getAllTimers: async () => {
        try {
            const response = await axiosInstance.get(`${BASE_URL}/getAll`);
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    getTimerById: async (timerID) => {
        try {
            const response = await axiosInstance.get(`${BASE_URL}/get/${timerID}`);
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    createTimer: async (timerData) => {
        try {
            const response = await axiosInstance.post(`${BASE_URL}/create`, timerData);
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    updateTimer: async (timerID, timerData) => {
        try {
            const response = await axiosInstance.put(`${BASE_URL}/update/${timerID}`, timerData);
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    deleteTimer: async (timerID) => {
        try {
            const response = await axiosInstance.delete(`${BASE_URL}/delete/${timerID}`);
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    restartTimer: async (timerID) => {
        try {
            const response = await axiosInstance.put(`${BASE_URL}/restart/${timerID}`);
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    getTimersByStudentId: async (studentId) => {
        try {
            const response = await axiosInstance.get(`${BASE_URL}/getByStudent/${studentId}`);
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    startTimer: async (timerID) => {
        try {
            const response = await axiosInstance.post(`${BASE_URL}/start/${timerID}`);
            return response.data;
        } catch (error) {
            throw error;
        }
    }
};

export default timerService;
